### ISSUE TITLE
feat: A way to know how many files are incoming in middleware

### ISSUE THREAD

**Issue by kwerdna19** [pingdotgg/uploadthing#565]
### Describe the feature you'd like to request

Please correct me if Im wrong but I dont think there is a way to tell how many files are incoming through a router middleware.

This is the use case I had in mind- on this app a user can upload a maximum of 5 images ever (at different times) - say they upload 3 images one day. And the next day they try to upload 3 more images, in the middleware I would like to lookup the users upload count and see if this count+incoming image count is less than 5, and if not, throw an error

For this, I would need the incoming image count somehow.

### Describe the solution you'd like to see

A parameter in the middleware func that tells us how many files are being processed

### Additional information

The same use case can be improved on the frontend too, as it seems there is no way to dynamically specify how many files are allowed to be dropped into a UTDropZone for example. But I think this can be left for a separate feature request / future consideration

### 👨‍👧‍👦 Contributing

- [x] 🙋‍♂️ Yes, I'd be down to file a PR implementing this feature!

**Comment by kwerdna19** [comment]
On looking at this further- it seems although we have access to the NextRequest, we cannot call `req.json()` because it has already been consumed by the UT internals. Attempting to call `req.clone()` seems to error out as well, for a reason Im not sure of.

But regardless, I think the fix is to pass the body back to the middleware after it is read the first time. That does raise the question of how to type this- since it is an arbitrarily parsed json body, its type is technically `any`. Should we use zod to ensure the type of this body ?

I can contribute a PR, just not sure exactly what the preference would be on that last question

**Comment by markflorkowski** [comment]
The way I would currently do what you are describing is using the `input` option to pass the number of files along to middleware, would that work for you?

Relevant docs:
- Server: https://docs.uploadthing.com/api-reference/server#input-1
- Client: https://docs.uploadthing.com/api-reference/react#configuration

**Comment by kwerdna19** [comment]
> The way I would currently do what you are describing is using the `input` option to pass the number of files along to middleware, would that work for you?
> 
> Relevant docs:
> 
> * Server: https://docs.uploadthing.com/api-reference/server#input-1
> * Client: https://docs.uploadthing.com/api-reference/react#configuration

Hello, thanks for the workaround suggestion - it seems valid, albeit maybe needlessly more code, but a couple things.

How would this work in practice. I am trying it out right now and having trouble.

Something like this on the frontend- is not working for me, presumably because uploadCount is updated too late (?) or at least not in time for the initial request -- on the middleware side input.count it yielding 0 for me

```
<UploadDropzone
        endpoint="attachments"
        onBeforeUploadBegin={files => {
          setUploadCount(files.length)
          return files
        }}
        input={{ count: uploadCount }}
      />
```

And 2nd question - is that method abusable? I.e. could a nefarious user change the input on the POST req? I suppose a user could change the upload input in general so this one I am less sure about.

**Comment by markflorkowski** [comment]
> How would this work in practice. I am trying it out right now and having trouble.

I think that using the provided components, there is just not enough granular control to manage this, since we don't expose the state, so you would need to use a custom component where you track the state yourself, and our `useUploadthing()` hook. I am looking into how we can better support your original ask for getting some file info in the `.middleware()` function

> presumably because uploadCount is updated too late

Yeah, I think that is what is happening here. `onBeforeUploadBegin` runs after middleware, before upload on the client. 

> could a nefarious user change the input on the POST req

yes, this is a valid concern. as you correctly identified, they could change the input in general as well, but since this is extra metadata, it's a bit harder to validate.

**Comment by kwerdna19** [comment]
Thanks @markflorkowski - appreciate the back and forth and you looking into it.

**Comment by juliusmarminge** [comment]
Will be released in 6.3
